clippy_lints/src/absolute_symbol_paths.rs

```diff
@@ -0,0 +1,119 @@
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::Span;
+use std::iter::once;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of symbols through absolute paths, like `std::env::current_dir`.
+    ///
+    /// ### Why is this bad?
+    /// Many codebases have their own style when it comes to symbol importing, but one that is
+    /// seldom used is using absolute paths *everywhere*. This is generally considered unidiomatic,
+    /// and you should add a `use` statement.
+    ///
+    /// Note: One exception to this is code from macro expansion - this does not lint such cases, as
+    /// using absolute paths is the proper way of referencing symbols in one.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let x = std::f64::consts::PI;
+    /// ```
+    /// Use any of the below instead, or anything else:
+    /// ```rust
+    /// # use std::f64;
+    /// # use std::f64::consts;
+    /// # use std::f64::consts::PI;
+    /// let x = f64::consts::PI;
+    /// let x = consts::PI;
+    /// let x = PI;
+    /// use std::f64::consts as f64_consts;
+    /// let x = f64_consts::PI;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub ABSOLUTE_SYMBOL_PATHS,
+    style,
+    "checks for usage of a symbol without a `use` statement"
+}
+impl_lint_pass!(AbsoluteSymbolPaths => [ABSOLUTE_SYMBOL_PATHS]);
+
+pub struct AbsoluteSymbolPaths {
+    pub absolute_symbol_paths_max_segments: u64,
+    pub absolute_symbol_paths_allow_std: bool,
+}
+
+impl LateLintPass<'_> for AbsoluteSymbolPaths {
+    fn check_crate(&mut self, cx: &LateContext<'_>) {
+        let Self {
+            absolute_symbol_paths_max_segments,
+            absolute_symbol_paths_allow_std,
+        } = *self;
+
+        cx.tcx.hir().visit_all_item_likes_in_crate(&mut V {
+            cx,
+            absolute_symbol_paths_max_segments,
+            absolute_symbol_paths_allow_std,
+        });
+    }
+}
+
+struct V<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    absolute_symbol_paths_max_segments: u64,
+    absolute_symbol_paths_allow_std: bool,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for V<'a, 'tcx> {
+    type NestedFilter = OnlyBodies;
+
+    fn nested_visit_map(&mut self) -> Self::Map {
+        self.cx.tcx.hir()
+    }
+
+    #[expect(clippy::cast_possible_truncation)]
+    fn visit_qpath(&mut self, qpath: &'tcx QPath<'tcx>, hir_id: HirId, span: Span) {
+        let Self {
+            cx,
+            absolute_symbol_paths_max_segments,
+            absolute_symbol_paths_allow_std,
+        } = *self;
+
+        if !span.from_expansion()
+            && let QPath::Resolved(_, path) = qpath
+            && let Some(def_id) = path.res.opt_def_id()
+            && let def_path = cx.tcx.def_path(def_id)
+            && let crate_name = cx.tcx.crate_name(def_path.krate)
+            && let segments = once(crate_name)
+                .chain(def_path.data.iter().filter_map(|segment| segment.data.get_opt_name()))
```

Comment:
I'll look into this soon

Reply:
This unfortunately won't work for something like `a::b::c::C`, where `a` is a module defined in the same crate. Makes sense, as `a` in this case isn't the crate root (but should still be linted)

Reply:
Why should that be linted? It isn't an absolute path

Reply:
While it's not "absolute" in the literal sense, there's still no use statement. Not linting it makes the lint ignore anything defined in the current crate, which isn't my favorite. This *could* ignore anything from the current crate intentionally but I'd much prefer also linting that

PS, it also ignores `crate::a::b::c::C` as `crate` has `Res::Err`.

Reply:
You can only refer to `crate::a::b::c::C` as `a::b::c::C` if you're in the root module because that's where `a` is defined or you have imported `a`, I don't think it makes sense to consider this an absolute path

Ah, in that case it might need to be something like

```rust
&& let Some(first_segment) = path.segments.first()
&& let Res::Def(DefKind::Mod, def_id) == first_segment.res
&& (def_id.index == CRATE_DEF_INDEX || first_segment.ident.name == kw::Crate)
```

Reply:
imo if you're referencing it without a `use` statement it should be linted, as `a::b::c::C` is still not too great. This is also a bit more obvious with the alternative suggested name for this lint: `symbol_reference_without_use`. Maybe we can rename it to something else to make it a bit clearer

Reply:
I think `a::b::c::C` is good enough. It helps with explicitness.
If we decided to lint both `crate::a::b::c::C` and `a::b::c::C`, the lint should be named something like `explicit_paths`, and I'm not sure it should be warn by default.

As a first iteration, I think only linting `crate::` is good enough, as it's the definition of an absolute path.

Reply:
It helps with explicitness, but if each segment isn't only a single character long then it'll take up half the screen. rustfmt or others may also decide to include a newline after the `=`, which doesn't help with readability in this case.

> and I'm not sure it should be warn by default.

It shouldn't be regardless ^^

> As a first iteration, I think only linting crate:: is good enough, as it's the definition of an absolute path.

You brought up an interesting idea in today's meeting (that I missed). In particular, linting inconsistent paths. I think linting only `crate` and external crates is a good idea if that's a separate lint since that *kind of* prevents that regardless